diff --git a/src/CollisionManager.cpp b/src/CollisionManager.cpp
index bffd8dd..96adab1 100644
--- a/src/CollisionManager.cpp
+++ b/src/CollisionManager.cpp
@@ -61,6 +61,7 @@ namespace mars
         void CollisionManager::itemAdded(const envire::core::TypedItemAddedEvent<envire::core::Item<interfaces::ContactPluginInterfaceItem>>& event)
         {
             auto& item = event.item->getData();
+            item.contactPluginInterface->setFrameID(event.frame);
             contactPluginItems.push_back(item);
         }
 
